### PR TITLE
[Translation][Extractor] Allow extracting an array of files besides extracting a directory

### PR DESCRIPTION
**Pull request by marcosdsanchez**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | N/A |
| License | MIT |
| Doc PR | N/A |

Give ability to extractors to extract from an array of files a file, or a Traversable besides a directory.

The following commit adds the new feature to the following extractors: 
PhpExtractor and TwigExtractor.
It also corrects the interface documentation to show the new options allowed as a parameter.

I found this change useful as I've got a custom translation:update command that instead of extracting from $rootPath. '/Resources/views/' it extracts from the files that git is showing as modified. The command usually runs much faster than the default as it only parses the needed files.

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Translation/TwigExtractor.php [hunk 1]
```diff
@@ -52,16 +52,22 @@ public function __construct(\Twig_Environment $twig)
...
      * {@inheritdoc}
      */
-    public function extract($directory, MessageCatalogue $catalogue)
+    public function extract($resource, MessageCatalogue $catalogue)
     {
-        // load any existing translation files
-        $finder = new Finder();
-        $files = $finder->files()->name('*.twig')->sortByName()->in($directory);
+        if (is_array($resource)) {
+            $files = array_filter($resource, function ($filename) {
```
Comment (aitboudad) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]:
```php
$files = array();
foreach ($resource as $file) {
    if (!is_file($file)) {
        throw new \InvalidArgumentException(sprintf('The "%s" file does not exist.', $file));
    }
        
    if ('twig' === pathinfo($filename, PATHINFO_EXTENSION)) {
        throw new \InvalidArgumentException('the given file type is not supported');
    }

    $files[] = new \SplFileInfo($file);
}
```

File: src/Symfony/Bridge/Twig/Translation/TwigExtractor.php [hunk 2]
```diff
@@ -52,16 +52,22 @@ public function __construct(\Twig_Environment $twig)
...
+                return 'twig' === pathinfo($filename, PATHINFO_EXTENSION);
+            });
+        } else {
+            $finder = new Finder();
+            $files = $finder->files()->name('*.twig')->in($resource);
+        }
         foreach ($files as $file) {
             try {
-                $this->extractTemplate(file_get_contents($file->getPathname()), $catalogue);
```
Comment (aitboudad) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: we can keep this as it

File: src/Symfony/Bridge/Twig/Translation/TwigExtractor.php (hunk 1 above)
Comment (marcosdsanchez) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: Some concerns about your comment:

- Isn't the `is_file` check too expensive if the array contains a lot of files?
- The second if statement is making this behave differently than before. Files with another extensions were ignored, no exception was thrown. I'd prefer to keep the old behaviour.
- Are we creating SplFileInfo instances only to keep the code from below the same?

Comment (aitboudad) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: 1. Anyway we need to be sure that file exist, the Finder is  expensive too
2. ok
3. Yes, it keep the code from below the same.

Comment (marcosdsanchez) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: Comments addressed. Also added more tests and different options for `$resource` like support for `Traversable`, a string containing a file, etc. Updated interface documentation too.

**Comment by marcosdsanchez** [comment]
Ci failure is not related.

**Comment by aitboudad** [comment]
@marcosdsanchez what do you think about adding a new class that give ability to extract from one file (FileTwigExtractor for example) ? /cc @fabpot

**Comment by marcosdsanchez** [comment]
@aitboudad , what about something like 1b419cb ?

**Comment by aitboudad** [comment]
@marcosdsanchez It's exactly what I want :D

File: src/Symfony/Bridge/Twig/Translation/TwigExtractor.php [hunk 3]
```diff
@@ -44,6 +45,11 @@ class TwigExtractor implements ExtractorInterface
      */
     private $twig;
 
+    /**
+     * @var string
+     */
+    private static $extension = 'twig';
```
Comment (aitboudad) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: we can avoid static here.

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 4]
```diff
@@ -52,13 +53,16 @@ class PhpExtractor implements ExtractorInterface
     );
 
     /**
+     * @var string
+     */
+    private static $extension = 'php';
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: same here

File: src/Symfony/Bridge/Twig/Translation/TwigExtractor.php (hunk 3 above)
Comment (marcosdsanchez) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: Should it be a constant instead?

Comment (aitboudad) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: something like https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Translation/Dumper/FileDumper.php#L104

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 4 above)
Comment (marcosdsanchez) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: Adressed comments. Rebased and squashed commits.

**Comment by aitboudad** [comment]
:+1:

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 5]
```diff
@@ -20,7 +21,7 @@
  *
  * @author Michel Salib <michelsalib@hotmail.com>
  */
-class PhpExtractor implements ExtractorInterface
+class PhpExtractor extends AbstractFileExtractor implements ExtractorInterface
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: Does it make sense to let `AbstractFileExtractor` implement `ExtractorInterface` instead? Or can we think about a use case where this is not necessary?

**Comment by marcosdsanchez** [comment]
@aitboudad  Because of the new class the code is not backwards compatible anymore. See https://travis-ci.org/symfony/symfony/builds/55525434 . What should I do? Get back to the version I had before or point the PR to another branch?

**Comment by aitboudad** [comment]
just update [composer.json](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Bridge/Twig/composer.json#L31) to 2.7

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php (hunk 5 above)
Comment (marcosdsanchez) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: I thought about this too but decided to go this way as the abstract class doesn't implement any of the methods from the interface.

**Comment by xabbuh** [comment]
@marcosdsanchez You will have to bump the version in the FrameworkBundle too.

**Comment by marcosdsanchez** [comment]
@aitboudad @xabbuh updated Twig Bridge and Framework Bundle. Travis is still failing with high dependencies in php 5.6 https://travis-ci.org/symfony/symfony/builds/55551337 . What am I missing?

**Comment by xabbuh** [comment]
@marcosdsanchez This seems to be related to how that specific build works (you change is not available in the components repos as it was not merged). So that's nothing to worry about.

**Comment by marcosdsanchez** [comment]
Thanks @xabbuh . I think this is ready then.

**Comment by xabbuh** [comment]
Can you please add an entry to the changelog?

**Comment by marcosdsanchez** [comment]
@xabbuh sure. For which version 2.3.27 (unreleased) ?

**Comment by fabpot** [comment]
@marcosdsanchez You need to add an entry in the CHANGELOG-2.7.md file.

**Comment by xabbuh** [comment]
@marcosdsanchez As @fabpot said, and the version should be 2.7.0.

**Comment by marcosdsanchez** [comment]
Thanks @fabpot and @xabbuh . Done in 5258f01.

File: src/Symfony/Bridge/Twig/Translation/TwigExtractor.php [hunk 6]
```diff
@@ -44,6 +45,11 @@ class TwigExtractor implements ExtractorInterface
      */
     private $twig;
 
+    /**
+     * @var string
+     */
+    private $extension = 'twig';
```
Comment (fabpot) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: I would remove this as there is no way to change it and no need to change it.

File: src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php [hunk 7]
```diff
@@ -52,13 +53,16 @@ class PhpExtractor implements ExtractorInterface
     );
 
     /**
+     * @var string
+     */
+    private $extension = 'php';
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Translation/PhpExtractor.php]: same as above.

**Comment by fabpot** [comment]
:+1: when my 2 comments are fixed

File: src/Symfony/Bridge/Twig/Translation/TwigExtractor.php (hunk 6 above)
Comment (marcosdsanchez) [inline src/Symfony/Bridge/Twig/Translation/TwigExtractor.php]: This was already discussed [here](https://github.com/symfony/symfony/pull/14002#discussion-diff-26958271R51). I'd prefer to have a class constant, but ok, I'll remove...

**Comment by marcosdsanchez** [comment]
@fabpot comments addressed.

**Comment by fabpot** [comment]
Thank you @marcosdsanchez.

**Pull request closed without merging**